lexicon/providers/dreamhost.py

```diff
@@ -0,0 +1,181 @@
+        super(Provider, self).__init__(config)
+        self.domain_id = None
+        self.api_endpoint = 'https://api.dreamhost.com/'
+
+    def _authenticate(self):
+        self.domain_id = None
+        payload = self._get('domain-list_domains')
+        data = payload.get('data', None)
+        if data is None:
+            raise Exception('Domain not found')
+
+        for domain in data:
+            if domain.get('domain', '') == self.domain:
+                self.domain_id = self.domain
+        if self.domain_id is None:
+            raise Exception('Domain not found')
+
+    def _create_record(self, rtype, name, content):
+        name = self._full_name(name)
+        payload = self._get('dns-add_record', query_params={
+            'record': name,
+            'type': rtype,
+            'value': content,
+        })
+
+        if (payload.get('result', '') != 'success' and
+                payload.get('data', '') != 'record_already_exists_remove_first' and
+                payload.get('data', '') != 'record_already_exists_not_editable'):
+            raise Exception('unable to add record: %s' % payload)
+
+        return True
+
+    # List all records. Return an empty list if no records found
+    # type, name and content are used to filter records.
+    # If possible filter during the query, otherwise filter after response is received.
+    def _list_records(self, rtype=None, name=None, content=None):
+
+        payload = self._get('dns-list_records')
+
+        if payload.get('result', '') != 'success':
+            raise Exception('unable to get records: %s' % payload)
+
+        resource_list = payload.get('data', None)
+        if resource_list is None:
+            raise Exception('unable to get records: %s' % payload)
+
+        resource_list = [
+            resource for resource in resource_list if resource['zone'] == self.domain]
+        if rtype:
+            resource_list = [
+                resource for resource in resource_list if resource['type'] == rtype]
+        if name:
+            name = self._full_name(name)
+            resource_list = [
+                resource for resource in resource_list if resource['record'] == name]
+        if content:
+            resource_list = [
+                resource for resource in resource_list if resource['value'] == content]
+
+        processed_records = []
+        for resource in resource_list:
+            processed_records.append({
+                'id': resource['type'] + '|' + resource['record'] + '|' + resource['value'],
+                'type': resource['type'],
+                'name': resource['record'],
+                'content': resource['value'],
+            })
+
+        return processed_records
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        if identifier == '':
+            identifier = None
+        if identifier is not None:
+            id_list = identifier.split('|')
```

Comment:
I see your smart move here, that avoids to request again the server and directly get the values you want from the identifier.

However, I would prefer a "real" ID (meaning something more like a hash, like in the `_identifier` I mentionned above, for two reasons:
* from a development point of view, this is brittle: for instance, if you have a `|` in your TXT content, you go in the gray zone with potential side-effects
* from a user point of view, this kind of concatenation will give not very handful ID to manipulate. For instance, a very long TXT content will give a very long ID. Carriage returns in a TXT value or weird characters will make it worse.

Reply:
The problem with dreamhost is that they provide a very primitive "dns-list_records" api
in that this api just dumps all the records ACROSS ALL DOMAINS that you own.
It means that it's not efficient to just get the list of records and do the filtering.

When deleting a record, dreamhost also requires that we need to provide the value to delete.
It means that we do need the value-information to let the users be able to update / delete the record.

I admit that "|" separation is a very bad implementation.
Though it's getting longer, I think I'll wrap the id with json.dumps({"type": type, "record": record, "value": value}).